contracts/extensions/ReputationBootstrapper.sol

```diff
@@ -0,0 +1,219 @@
+      require(_amounts[i] <= INT128_MAX, "reputation-bootstrapper-invalid-amount");
+
+      if (_paid[i]) {
+        uint256 priorGrant = grants[_paid[i]][_hashedSecrets[i]].amount;
+        totalPayableGrants = totalPayableGrants - priorGrant + _amounts[i];
+      }
+
+      grants[_paid[i]][_hashedSecrets[i]] = Grant(_amounts[i], block.timestamp);
+
+      emit GrantSet(_paid[i], _hashedSecrets[i], _amounts[i]);
+    }
+
+    require(totalPayableGrants <= balance, "reputation-bootstrapper-insufficient-balance");
+  }
+
+  /// @notice Commit the secret, beginning the security delay window
+  /// @param _committedSecret A sha256 hash of (userAddress, secret)
+  function commitSecret(bytes32 _committedSecret) public notDeprecated {
+    bytes32 addressHash = keccak256(abi.encodePacked(msgSender(), _committedSecret));
+    committedSecrets[addressHash] = block.timestamp;
+  }
+
+  /// @notice Claim the grant, after committing the secret and having the security delay elapse
+  /// @param _secret The secret corresponding to a reputation grant
+  function claimGrant(bool _paid, uint256 _secret) public notDeprecated {
+    bytes32 committedSecret = keccak256(abi.encodePacked(msgSender(), _secret));
+    bytes32 addressHash = keccak256(abi.encodePacked(msgSender(), committedSecret));
+    uint256 commitTimestamp = committedSecrets[addressHash];
+
+    require(
+      commitTimestamp > 0 && commitTimestamp + SECURITY_DELAY <= block.timestamp,
+      "reputation-bootstrapper-commit-window-unelapsed"
+    );
+
+    bytes32 hashedSecret = keccak256(abi.encodePacked(_secret));
+    uint256 grantAmount = grants[_paid][hashedSecret].amount;
+    uint256 grantTimestamp = grants[_paid][hashedSecret].timestamp;
+
+    require(grantAmount > 0, "reputation-bootstrapper-nothing-to-claim");
+
+    delete grants[_paid][hashedSecret];
+
+    if (_paid) {
+      totalPayableGrants -= grantAmount;
+      require(ERC20(token).transfer(msgSender(), grantAmount), "reputation-bootstrapper-transfer-failed");
+    }
+
+    uint256 decayEpochs = (block.timestamp - grantTimestamp) / decayPeriod;
+    uint256 adjustedNumerator = decayNumerator;
+
+    // This algorithm successively doubles the decay factor while halving the number of epochs
+    // This allows us to perform the decay in O(log(n)) time
+    // For example, a decay of 50 epochs would be applied as (k**2)(k**16)(k**32)
+    while (decayEpochs > 0){
+      // slither-disable-next-line weak-prng
+      if (decayEpochs % 2 >= 1) {
+        // slither-disable-next-line divide-before-multiply
+        grantAmount = grantAmount * adjustedNumerator / decayDenominator;
+      }
+      // slither-disable-next-line divide-before-multiply
+      adjustedNumerator = adjustedNumerator * adjustedNumerator / decayDenominator;
+      decayEpochs >>= 1;
+    }
+
+    colony.emitDomainReputationReward(1, msgSender(), int256(grantAmount));
+
+    emit GrantClaimed(msgSender(), grantAmount, _paid);
```

Comment:
Have we (or product?) given any amount of consideration as to which `grantAmount` should actually be emitted here (i.e. the decayed amount, or the original)?

Reply:
Not to my knowledge.

@arrenv, do you have any preference? When a user claims a grant, should we emit the initial amount, the decayed amount actually claimed, or both?

Reply:
It makes the most sense that this should emit the decayed amount, as they never receive the initial amount and thus it would be inaccurate.